Add EdgeDB Auth + Express example app

scotttrinh opened the pull request:
Very barebones example app:

- adds the required auth-related routes with idiomatic Express route handling
  - Show success handling with route handlers
  - Show error handling with route-specific error handlers
- shows how to create middleware for protecting routes
- shows using the session middleware to add an authenticated client

Blocked on publishing the express auth helper package, but tested locally via `npm link`.

raddevon reviewed (commented):
This is looking great! Left a couple of comments about comments. 😜 Generally, I think we should make it very clear what we're doing anytime we include code that is not actually being used by this app but is meant as an example, to make why it's there and what purpose it serves unambiguous.

I love the idea you proposed on Slack of adding some API routes and showing them making authenticated queries. Feels like the natural next question from someone who is using this example app to learn EdgeDB Auth.

scotttrinh commented on express-auth/auth.ts:
```diff
@@ -0,0 +1,186 @@
+  callback: [builtinCallback],
+});
+
+export const builtinUIRouter = Router()
```
Actually, I wonder if I should demonstrate both, since the manual router needs to mounted to a route since it doesn't define a route itself. Maybe I can make two apps, one that uses the manual router and has a custom UI, and one that uses the built-in UI with the factory-style router?

That would address both of these "dead-code" suggestions. WDYT?

raddevon replied:
Yes! I like the idea of helping people connect those dots.

scotttrinh replied:
⚡ @raddevon https://github.com/edgedb/edgedb-examples/pull/116/files/1e78110c50e05ff38a17f3f0d7c1e26313136281..fec57be36a6e1699186a12e4b2631311dfccea96

raddevon commented:
We should add some step-by-step instructions to the readme on how to run this app. I haven't been able to do that successfully myself, and I'm not sure if that is due to the express auth helper package you mentioned in your description or due to something else. Here's what I tried:

1. Run `edgedb project init`
2. Run `npm install`
3. Run `npm run dev`

I get this:
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/raddevon/Documents/projects/edgedb/edgedb-examples/express-auth/index.ts' imported from /Users/raddevon/Documents/projects/edgedb/edgedb-examples/express-auth/
```

scotttrinh commented:
@raddevon 

Thanks for the reminder, I actually fixed this already but didn't update the version of the library installed here. Will fix.

scotttrinh commented:
@raddevon ready to review!

raddevon commented on express-auth/README.md:
````diff
@@ -0,0 +1,13 @@
+- `builtin.ts` shows how to use the built-in UI using the route factory method.
+- `custom.ts` shows how to build your own UI using the manual router method.
+
+## Usage
+
+```bash
+npm install
+npm run builtin
+# or
+npm run custom
+```
````
It took me a bit to find exactly what these settings needed to be, and I think it might be worth including them in the readme.
```suggestion
## Setup

1. Run `edgedb project init`and follow the prompts
2. Run `npm install`
3. Run `edgedb ui`
4. Switch to the "Auth Admin" section (nested shield icon in the left-hand toolbar)
5. Enter a value for `auth_signing_key` or click the button nested in the field to generate one and click the "Update" button
6. Enter `http://localhost:3333/` in the `allowed_redirect_urls` field and click "Update"
7. Switch to the "Providers" tab
8. Click "Add Provider," select and configure a provider, and click "Add Provider" to save the new provider
9. Set the value for `redirect_to` to `http://localhost:3333/auth/callback` and click "Update"
10. (Built-in only) Click "Enable" under "Login UI"

> Note: If you configured the "Email + Password" provider to require verification, the example app will need a way to send emails. You might try running Mailpit to do this for local testing. We share how [in our EdgeDB Auth guide](https://www.edgedb.com/docs/guides/auth/index#email-and-password).

## Usage

```bash
npm run builtin
# or
npm run custom
```

Still trying to make the custom UI work, so I may return to this and update it.

raddevon reviewed (commented):
I'm having some trouble getting the custom UI to work. I tinkered a bit and figure out how to get the built-in UI to work, and I've captured that process as a suggestion on the readme for the example project. Once I can figure out the custom UI, I may need to circle back and update that suggestion depending on whether the process is different. My goal is that the readme would be able to take a user from zero to hero and get the app fully running in either permutation they want to try.

At the moment, when I try to sign up using the custom UI and the email/password provider, I get "Cannot POST /auth/signup" in the browser. Here's the server console output:
```
Listening on http://localhost:3333
  express:router dispatching GET /signin +7s
  express:router query  : /signin +0ms
  express:router expressInit  : /signin +1ms
  express:router jsonParser  : /signin +0ms
  body-parser:json skip empty body +0ms
  express:router urlencodedParser  : /signin +0ms
  body-parser:urlencoded skip empty body +1ms
  express:router cookieParser  : /signin +0ms
  express:router <anonymous>  : /signin +0ms
  express:router dispatching GET /signin +5s
  express:router query  : /signin +0ms
  express:router expressInit  : /signin +0ms
  express:router jsonParser  : /signin +1ms
  body-parser:json skip empty body +0ms
  express:router urlencodedParser  : /signin +0ms
  body-parser:urlencoded skip empty body +0ms
  express:router cookieParser  : /signin +0ms
  express:router <anonymous>  : /signin +0ms
  express:router dispatching POST /auth/signup +5s
  express:router query  : /auth/signup +1ms
  express:router expressInit  : /auth/signup +0ms
  express:router jsonParser  : /auth/signup +0ms
  body-parser:json content-type "application/x-www-form-urlencoded" +1ms
  body-parser:json skip parsing +2ms
  express:router urlencodedParser  : /auth/signup +0ms
  body-parser:urlencoded content-type "application/x-www-form-urlencoded" +0ms
  body-parser:urlencoded content-encoding "identity" +1ms
  body-parser:urlencoded read body +0ms
  body-parser:urlencoded parse body +16ms
  body-parser:urlencoded parse urlencoding +0ms
  express:router cookieParser  : /auth/signup +1ms
  express:router <anonymous>  : /auth/signup +0ms
  finalhandler default 404 +0ms
```

Have I goofed here, or is something broken? I tried changing the action on the form to `/auth/email-password/signup`, which seemed to get me a bit further, but I just got a different 404 so I figured I should stop before I really break something. 😅

scotttrinh commented:
@raddevon 

Apologies for the sloppy implementation of the separate "custom UI" app. I had it working when I was switching back and forth, but when I moved it to a separate app entrypoint, I didn't do any testing and used a different route structure, as you noticed.

You were correct to change the actions to have an `email-password` prefix route, but the other thing I wasn't handling here was if you had `require_verification` set to true 🤦 I personally never use that option but didn't think about how it would affect the example. I've fixed that now as well!

I'll merge you README suggestion and feel free to check it out again. Thanks for the help and apologies for having to go back and forth here!

raddevon commented:
No need for apologies, @scotttrinh. That back and forth is exactly what review is for!

Thank you! I'll give it another look.

raddevon commented on express-auth/README.md:
```diff
@@ -0,0 +1,26 @@
+1. Run `edgedb project init`and follow the prompts
+2. Run `npm install`
+3. Run `edgedb ui`
+4. Switch to the "Auth Admin" section (nested shield icon in the left-hand toolbar)
```
```suggestion
4. Click into the `edgedb` database and switch to the "Auth Admin" section (nested shield icon in the left-hand toolbar)
```

raddevon commented on express-auth/README.md:
```diff
@@ -0,0 +1,26 @@
+6. Enter `http://localhost:3333/` in the `allowed_redirect_urls` field and click "Update"
+7. Switch to the "Providers" tab
+8. Click "Add Provider," select and configure a provider, and click "Add Provider" to save the new provider
+9. Set the value for `redirect_to` to `http://localhost:3333/auth/callback` and click "Update"
+10. (Built-in only) Click "Enable" under "Login UI"
```
```suggestion
9. (Built-in only) Click "Enable" under "Login UI"
10. (Built-in only) Set the value for `redirect_to` to `http://localhost:3333/auth/callback` and click "Update"
```

raddevon commented on express-auth/auth.ts:
```diff
@@ -0,0 +1,191 @@
+  .post("/signin", auth.emailPassword.signIn, ...emailPassword.signIn)
+  .post(
+    "/signup",
+    auth.emailPassword.signUp("http://localhost:3333/auth/verify"),
```
```suggestion
    auth.emailPassword.signUp("http://localhost:3333/auth/email-password/verify"),
```

raddevon reviewed (approved):
I think we're good to merge after you take a look at these suggestions. I found a couple of mistakes on my part in the setup instructions. Then I found one more place where we need the `email-password` segment: in the verification link in the email.
